bazel/linux/qemu.bzl

```diff
@@ -0,0 +1,120 @@
+load("//bazel/linux:runner.bzl", "CREATE_RUNNER_ATTRS", "create_runner")
+load("@bazel_skylib//lib:shell.bzl", "shell")
+
+DEFAULT_QEMU_FLAGS = [
+    "-enable-kvm",
+    "-cpu",
+    "host",
+    "-machine",
+    "pc,accel=kvm,usb=off,dump-guest-core=off",
+    "-m",
+    "2048",
+    "-smp",
+    "4,sockets=4,cores=1,threads=1",
+    "-rtc",
+    "base=utc",
+    "-boot",
+    "strict=on",  # Only boot from devices specified, in the order specified.
+    "-nographic",  # No UI, and stdio acts as ttyS0.
+    "-no-reboot",  # When kernel crashes / exits, don't restart VM.
+]
+
+DEFAULT_KERNEL_FLAGS = [
+    "rw",
+    "console=ttyS0",  # early boot messages to serial, which qemu sends to stdio.
+    "panic=-1",  # reboot (exit qemu) immediately on panic.
+]
+
+def _kernel_qemu_test(ctx):
+    code = """
+QEMU_FLAGS={qemu_flags}
+QEMU_SEARCH={qemu_search}
+KERNEL_FLAGS={kernel_flags}
+
+for qemu in "${{QEMU_SEARCH[@]}}"; do
+    QEMU_BINARY=$(which "$qemu") && break
+done
+test -n "$QEMU_BINARY" || {{
+    echo 1>&2 "NO QEMU FOUND - Looked in: ${{QEMU_SEARCH[@]}}"
+    echo 1>&2 "Use qemu_binary= or qemu_search= in the rule to tune the behavior"
+    exit 10
+}}
+
+if [ -n "$ROOTFS" ]; then
+    QEMU_FLAGS+=("-drive" "file=$ROOTFS,if=virtio,cache=none")
+else
+    QEMU_FLAGS+=("-fsdev" "local,security_model=none,multidevs=remap,id=fsdev-fsRoot,path=/")
+    QEMU_FLAGS+=("-device" "virtio-9p-pci,fsdev=fsdev-fsRoot,mount_tag=/dev/root")
+    KERNEL_FLAGS+=("root=/dev/root" "rootfstype=9p" "init=$INIT")
+    KERNEL_FLAGS+=("rootflags=trans=virtio,version=9p2000.L,msize=5000000,cache=mmap,posixacl")
+fi
+test -z "$KERNEL" || QEMU_FLAGS+=("-kernel" "$KERNEL")
+test -z "$INTERACTIVE" || KERNEL_FLAGS+=("init=/bin/sh")
+
+QEMU_FLAGS+=("-append" "${{KERNEL_FLAGS[*]}}")
+
+echo Running qemu: "$QEMU_BINARY" "${{QEMU_FLAGS[@]}}"
+if [ -z "$INTERACTIVE" ]; then
+    "$QEMU_BINARY" "${{QEMU_FLAGS[@]}}" | tee "$OUTPUTFILE"
+else
+    "$QEMU_BINARY" "${{QEMU_FLAGS[@]}}"
+fi
+"""
+    qemu_search = ctx.attr.qemu_search
+    runfiles = None
+    if ctx.attr.qemu_binary:
+        di = ctx.attr.qemu_binary[DefaultInfo]
+        qemu_search = [di.executable]
+        runfiles = di.runfiles
+    qemu_flags = ctx.attr.qemu_defaults + ctx.attr.qemu_flags
+
+    kernel_flags = ctx.attr.kernel_defaults + ctx.attr.kernel_flags
+    return create_runner(ctx, ctx.attr.archs, code, runfiles = runfiles, extra = {
+        "qemu_search": shell.array_literal(qemu_search),
+        "qemu_flags": shell.array_literal(qemu_flags),
+        "kernel_flags": shell.array_literal(kernel_flags),
+    })
+
+kernel_qemu_test = rule(
```

Comment:
kunit_qemu_test given the specific use case

Reply:
Let me update the comments, copied over from other code, good catch.

Tl;Dr: this is not tied with kunit at all: this just runs code in qemu, whatever that is, and saves the result.
An optional checker is used by the caller to verify the output.

Additionally ... with @george-enf we discussed renaming all those rules, as the names don't really add up anymore. But would rather do a separate cleanup sweep across the directory - so keeping the existing naming convention for now.

Reply:
added commit, marking as resolved.